File: test/test_botvel.py
import socket
import argparse


def encode_msg(v_l,v_r):

    max_vel = (1 << 10) - 1
    left_vel = min(abs(int(v_l)), max_vel)
    right_vel = min(abs(int(v_r)), max_vel)

    left_bits = format(left_vel, '010b')
    right_bits = format(right_vel, '010b')

    # print("LEFT BITS ",left_bits)

    bit_string = (f"1{right_bits}00000" if v_r < 0 else f"0{right_bits}00000") + (f"1{left_bits}00000" if v_l < 0 else f"0{left_bits}00000")
    # print("Bit string: ", bit_string)
    ascii_string = ""
    for i in range(0, len(bit_string), 8):
        chunk = bit_string[i:i + 8]
        # print("Chunk: ", chunk)
        ascii_string += chr(int(chunk, 2))
        # print("Ascii string: ", ascii_string)

    # print(f"BIT String: {bit_string}")

    return ascii_string

def test_botvel(port,ip,v_l,v_r):
    # Create a socket object
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

    # Send this encoded message to the robot
    i = 0

    while i<1000:
        s.sendto(encode_msg(v_l,v_r).encode('utf-8'), (ip, port))
        i+=1
        # print("Sent message")

    s.close()

if __name__ == "__main__":
    # Pass the port and ip address of the robot as arguments and take from command line

    parser = argparse.ArgumentParser()

    parser.add_argument("--port", help="Port number of the robot", type=int, default=5005)
    parser.add_argument("--ip", help="IP address of the robot", type=str, default="10.42.0.40")
    parser.add_argument("--v_l", help="Left wheel velocity", type=int, default=20)
    parser.add_argument("--v_r", help="Right wheel velocity", type=int, default=30)
    args = parser.parse_args()

    port = args.port
    ip = args.ip
    v_l = args.v_l
    v_r = args.v_r

    print("VL ",v_l)
    print("VR ",v_r)


    test_botvel(port,ip,v_l,v_r)
    
    

    

File: test/test_botvel.cpp
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <string>
#include<bits/stdc++.h>
#include<time.h>
#include<thread>
#include <chrono>
using namespace std;



int main(int argc,char *argv[]){

    int dgram_socket = socket(PF_INET, SOCK_DGRAM, 0);

    struct sockaddr_in dest;
    dest.sin_family = AF_INET;

    char* ip = argv[1];
    int port = atoi(argv[2]);

    inet_aton(ip, &dest.sin_addr);
    dest.sin_port = htons(port);

    int v_r = atoi(argv[3]);
    int v_l = atoi(argv[4]);

    cout<<"Bot Velocity(firacomm) "<<": "<<(int)v_l<<" "<<(int)v_r<<endl;
    const int max_vel = (1<<10) - 1;
    int leftvel = std::min(abs((int)v_l),max_vel);
    int rightvel = std::min(abs((int)v_r),max_vel);
    std::bitset<10> leftBits(leftvel);
    std::bitset<10> rightBits(rightvel);
    std::string bitString =  (v_l < 0 ? "1" : "0") + leftBits.to_string() + "00000" + (v_r < 0 ? "1" : "0") +  rightBits.to_string() + "00000";
    std::string asciiString;
    for (size_t i = 0; i < bitString.length(); i += 8) {
        std::bitset<8> chunk(bitString.substr(i, 8));
        asciiString += static_cast<char>(chunk.to_ulong());
    }
    std::cout << "Ascii String: " << bitString <<endl;

    while (true)
    {
        sendto(dgram_socket, asciiString.data(), asciiString.size(), 0, (struct sockaddr*)&dest, sizeof(dest));
        std::this_thread::sleep_for(std::chrono::milliseconds(3));
        // sleep(2);
    }
    

}